In [1]:
from sklearn.datasets import load_boston
import matplotlib.pyplot as plt

In [2]:
boston=load_boston()
print(boston.keys())
print(boston.items())

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])
dict_items([('data', array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])), ('target', array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6

In [3]:
x=boston.data
y=boston.target
len(x), len(y)

(506, 506)

## 데이터셋 나누기

In [4]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test=train_test_split(x,y,   #학습 75%, value data 25%
                                                  test_size=0.25,
                                               random_state=42)

## 모델 만들어 학습 및 예측

#### 1. 모델 만들기
#### 2. 모델 학습  model.fit()
#### 3. 모델 예측 model.predict()
#### 4. 모델 평가

#### 범주형일 때는 knn을 사용
#### 수치형일 때는 LinearRegression을 사용

In [5]:
from sklearn.linear_model import LinearRegression
model=LinearRegression().fit(x_train,y_train)
pred=model.predict(x_test)
pred

array([28.83885359, 36.00783288, 15.08324755, 25.23090886, 18.87864064,
       23.21398327, 17.5931124 , 14.30508093, 23.05438985, 20.62008346,
       24.78514683, 18.66833668, -6.9788951 , 21.83575737, 19.20898992,
       26.2868054 , 20.54379176,  5.65713224, 40.42358065, 17.64146116,
       27.32258958, 30.05056174, 11.15013704, 24.11530393, 17.89145648,
       15.79348591, 22.94743453, 14.2586068 , 22.26731194, 19.24709013,
       22.26897546, 25.24344002, 25.69165643, 17.98759507, 16.70286649,
       17.11631225, 31.19643534, 20.17835831, 23.71828436, 24.79196868,
       13.94575895, 32.00389982, 42.53869791, 17.44523722, 27.15354457,
       17.07482215, 13.89272021, 26.06440323, 20.36888769, 29.97813037,
       21.35346608, 34.32287916, 15.88498671, 26.17757739, 39.50970314,
       22.84123308, 18.95049088, 32.68913818, 25.02057949, 12.90539147,
       22.76052302, 30.53884316, 31.60797905, 15.92162168, 20.50670563,
       16.50798147, 20.50202198, 26.00723901, 30.63860954, 11.42

In [6]:
import pandas as pd
df=pd.DataFrame({"y_test" : y_test,"prediction" : pred, "diff" : (y_test - pred)})
df

y_test  prediction       diff
0      23.6   28.838854  -5.238854
1      32.4   36.007833  -3.607833
2      13.6   15.083248  -1.483248
3      22.8   25.230909  -2.430909
4      16.1   18.878641  -2.778641
5      20.0   23.213983  -3.213983
6      17.8   17.593112   0.206888
7      14.0   14.305081  -0.305081
8      19.6   23.054390  -3.454390
9      16.8   20.620083  -3.820083
10     21.5   24.785147  -3.285147
11     18.9   18.668337   0.231663
12      7.0   -6.978895  13.978895
13     21.2   21.835757  -0.635757
14     18.5   19.208990  -0.708990
15     29.8   26.286805   3.513195
16     18.8   20.543792  -1.743792
17     10.2    5.657132   4.542868
18     50.0   40.423581   9.576419
19     14.1   17.641461  -3.541461
20     25.2   27.322590  -2.122590
21     29.1   30.050562  -0.950562
22     12.7   11.150137   1.549863
23     22.4   24.115304  -1.715304
24     14.2   17.891456  -3.691456
25     13.8   15.793486  -1.993486
26     20.3   22.947435  -2.647435
27     14.9   14.258607   0.641393
28     21.7   22.267312  -0.567312
29     18.3   19.247090  -0.947090
..      ...         ...        ...
97     17.9   -0.774509  18.674509
98      9.6   13.338757  -3.738757
99     17.2   15.627032   1.572968
100    22.5   22.217554   0.282446
101    21.4   24.581887  -3.181887
102    12.0   10.773022   1.226978
103    19.9   19.500684   0.399316
104    19.4   23.234504  -3.834504
105    13.4   11.773888   1.626112
106    18.2   18.367779  -0.167779
107    24.6   25.438379  -0.838379
108    21.1   20.890792   0.209208
109    24.7   24.084406   0.615594
110     8.7    7.365872   1.334128
111    27.5   19.164243   8.335757
112    20.7   21.937341  -1.237341
113    36.2   27.411917   8.788083
114    31.6   32.508572  -0.908572
115    11.7   14.868852  -3.168852
116    39.8   35.059125   4.740875
117    13.9   12.860751   1.039249
118    21.8   20.830436   0.969564
119    23.7   28.420771  -4.720771
120    17.6   15.658537   1.941463
121    24.4   24.671964  -0.271964
122     8.8    3.284209   5.515791
123    19.2   23.798796  -4.598796
124    25.3   25.733299  -0.433299
125    20.4   23.048156  -2.648156
126    23.1   24.730468  -1.630468

[127 rows x 3 columns]

In [7]:
#MAE, MSE 가 낮을 수록 좋다.

MAE=sum(abs(df["diff"]))/len(df["diff"])
MSE=sum((df["diff"])**2)/len(df["diff"])

import numpy as np

RMSE=np.sqrt(MSE)

print(MAE)
print(MSE)
print(RMSE)

3.060939595437035
22.098694827098043
4.700924890603767


In [8]:
acc_tr=model.score(x_train,y_train)
acc_test=model.score(x_test,y_test)
print("훈련 세트의 정확도 : {:.2f}".format(acc_tr))  
print("테스트 세트의 정확도 : {:.2f}".format(acc_test))  

훈련 세트의 정확도 : 0.75
테스트 세트의 정확도 : 0.68


test_size=0.25
MAE=3.0609395954370586
MSE=22.0986948270982
RMSE=4.700924890603784
훈련 세트의 정확도 : 0.75
테스트 세트의 정확도 : 0.68

test_size=0.30
MAE=3.1627098714574227
MSE=21.517444231177397
RMSE=4.638689926172841
훈련 세트의 정확도 : 0.74
테스트 세트의 정확도 : 0.71

test_size=0.10
MAE=2.834210457858976
MSE=14.995852876582854
RMSE=3.872447917865759
훈련 세트의 정확도 : 0.74
테스트 세트의 정확도 : 0.76